src/neuroglancer/datasource/deepzoom/frontend.ts

```diff
@@ -0,0 +1,304 @@
+    const resolution = this.resolution = new Float64Array(rank);
+    const voxelOffset = this.voxelOffset = new Float32Array(rank);
+    const size = this.size = new Float32Array(rank);
+    if (rank === 4) {
+      resolution[3] = 1;
+      size[3] = numChannels;
+    }
+    verifyObjectProperty(
+        obj, 'resolution',
+        x => parseFixedLengthArray(resolution.subarray(0, 3), x, verifyFinitePositiveFloat));
+    verifyOptionalObjectProperty(
+        obj, 'voxel_offset', x => parseFixedLengthArray(voxelOffset.subarray(0, 3), x, verifyInt));
+    verifyObjectProperty(
+        obj, 'size', x => parseFixedLengthArray(size.subarray(0, 3), x, verifyPositiveInt));
+    this.chunkSizes = verifyObjectProperty(
+        obj, 'chunk_sizes', x => parseArray(x, y => {
+                              const chunkSize = new Uint32Array(rank);
+                              if (rank === 4) chunkSize[3] = numChannels;
+                              parseFixedLengthArray(chunkSize.subarray(0, 3), y, verifyPositiveInt);
+                              return chunkSize;
+                            }));
+    if (this.chunkSizes.length === 0) {
+      throw new Error('No chunk sizes specified.');
+    }
+    this.encoding =
+        verifyObjectProperty(obj, 'encoding', x => verifyEnumString(x, ImageTileEncoding));
+    this.key = verifyObjectProperty(obj, 'key', verifyString);
+  }
+}
+
+/*export*/ interface PyramidalImageInfo {
+  dataType: DataType;
+  volumeType: VolumeType;
+  scales: ScaleInfo[];
+  modelSpace: CoordinateSpace;
+  overlap: number;
+  tilesize: number;
+}
+
+/*export*/ function buildPyramidalImageInfo(metadata: DZIMetaData): PyramidalImageInfo {
+  const {width, height, tilesize, overlap, format} = metadata;
+  const dataType = DataType.UINT8;
+  const numChannels = 3;
+  const volumeType = VolumeType.IMAGE;
+  const scaleInfos = new Array<ScaleInfo>();
+  let w = width, h = height;
+  let maxlevel = Math.ceil(Math.log2(Math.max(w, h)));
+  do {
+    const lvl = scaleInfos.length;
+    const res = 1 << lvl;
+    scaleInfos.push(new ScaleInfo(
+        {
+          key: (maxlevel - lvl).toString(),
+          size: [w, h, 1],
+          resolution: [res, res, res],
+          chunk_sizes: [[tilesize, tilesize, 1]],
+          encoding: format
+        },
+        numChannels));
+    w = Math.ceil(w / 2);
+    h = Math.ceil(h / 2);
+  } while (w > 1 || h > 1);
+
+  if (scaleInfos.length === 0) throw new Error('Expected at least one scale');
+  const baseScale = scaleInfos[0];
+  const rank = 4;
+  const scales = new Float64Array(rank);
+  const lowerBounds = new Float64Array(rank);
+  const upperBounds = new Float64Array(rank);
+  const names = ['x', 'y', 'z'];
+  const units = ['m', 'm', 'm'];
+
+  for (let i = 0; i < 3; ++i) {
+    scales[i] = baseScale.resolution[i] / 1e9;
+    lowerBounds[i] = baseScale.voxelOffset[i];
+    upperBounds[i] = lowerBounds[i] + baseScale.size[i];
+  }
+  if (rank === 4) {
```

Comment:
This extra logic is not applicable here, right?

Reply:
Done in https://github.com/google/neuroglancer/commit/ceca06e04a75b5eb39eb435dbd3058decefb5d85